eravm-stable-interface/src/tracer_interface.rs

```diff
@@ -0,0 +1,255 @@
+use crate::StateInterface;
+
+macro_rules! forall_simple_opcodes {
+    ($m:ident) => {
+        $m!(Nop);
+        $m!(Add);
+        $m!(Sub);
+        $m!(And);
+        $m!(Or);
+        $m!(Xor);
+        $m!(ShiftLeft);
+        $m!(ShiftRight);
+        $m!(RotateLeft);
+        $m!(RotateRight);
+        $m!(Mul);
+        $m!(Div);
+        $m!(NearCall);
+        $m!(Jump);
+        $m!(Event);
+        $m!(L2ToL1Message);
+        $m!(Decommit);
+        $m!(This);
+        $m!(Caller);
+        $m!(CodeAddress);
+        $m!(ErgsLeft);
+        $m!(SP);
+        $m!(ContextMeta);
+        $m!(ContextU128);
+        $m!(SetContextU128);
+        $m!(IncrementTxNumber);
+        $m!(AuxMutating0);
+        $m!(PrecompileCall);
+        $m!(HeapRead);
+        $m!(HeapWrite);
+        $m!(AuxHeapRead);
+        $m!(AuxHeapWrite);
+        $m!(PointerRead);
+        $m!(PointerAdd);
+        $m!(PointerSub);
+        $m!(PointerPack);
+        $m!(PointerShrink);
+        $m!(StorageRead);
+        $m!(StorageWrite);
+        $m!(TransientStorageRead);
+        $m!(TransientStorageWrite);
+    };
+}
+
+macro_rules! pub_struct {
+    ($x:ident) => {
+        pub struct $x;
+    };
+}
+
+pub mod opcodes {
+    forall_simple_opcodes!(pub_struct);
+    pub struct FarCall<const M: u8>;
+    pub struct Ret<const M: u8>;
```

Comment:
I thought that this would require some ugly code on the VM side but it actually doesn't if I also switch the VM to this kind of solution. `fn ret<T: Tracer, const RETURN_TYPE: u8,` has the exact same problem where I'd like to have a const enum but that isn't allowed.

Reply:
This is now fixed. Never thought of dependent types as a way to avoid menial labor before 😄